### ISSUE TITLE
bem-te-vi challenge: project sketch

### ISSUE THREAD

**Issue by AlexandreHiroyuki** [AlexandreHiroyuki/mock-devboard#1]
## Goals

- Use the ULTIMATE bem-te-vi stack (Hono, NodeJS, TypeScript, PostgreSQL, Remix).
- Good UX.
- Document everything important. It's not only about the code but the reasoning behind every decision in the code.
- Finish in one week.
- Open Source Experience.
- Communication/Teaching.

- Learn how to use Remix.
- Learn how to use Hono.
- Learn how to use Panda CSS.
- Learn how to use React Aria.
- Learn how to use Zod.
- Learn how to use Swagger.

## Thematic Choice

An application closer to the insurance context would be preferable, but I lack knowledge about innovation in this market. My second choice is a context with which I'm more familiar because I can show more about my reasoning about problems and how I could act after learning more context when joining the company.

My last job and current college projects were about IoT systems (embedded and web development), therefore where I can speak the most with less time.

### Prerequisites
- Has a list.
- Use authentication.
- Create pages for:
  - Creating a new record (form-based interaction).
  - Reading existing records (display records in a list or table).
  - Updating existing records (editable form).
  - Deleting existing records (confirmation prompt before deletion).
- Data Fetching with Remix Loaders.
- Responsive front-end design.

### Conclusion

Because the presentation is coming up soon, it's better to treat this project as a Proof of Concept.

I will document everything in English since the codebase will be open source.

- Home page.
- Sign In / Sign Up.
- Dashboard with a list of devices.
- A specific device page that plots its data.

## Extra ideas

- Use an LLM to provide a periodic report of a group of devices.

**Comment by AlexandreHiroyuki** [comment]
# Dependencies Required by the Challenge

## Front-end

- Remix
- React
- TypeScript
- Panda CSS
- React Aria

## Back-End

- Porsager Postgres
- TypeScript
- NodeJS
- Hono
- Zod
- Swagger

## Database

- Postgres

## Extra

- Scalar

**Comment by AlexandreHiroyuki** [comment]
## Database

In every project I work on, I prefer to keep every dependency possible isolated in its project to avoid conflicts with different development environments and other codebases on the same machine.

With that in mind, I installed PostgreSQL through Docker instead of directly on my computer.

**Comment by AlexandreHiroyuki** [comment]
# About workspaces

The front-end and back-end share some dependencies, which makes me think about using workspaces to manage and standardize versions across these codes and share schemas.

However, for a one-week project, I considered a bit of over-engineering because I could not risk losing time to environment setup issues.

**Comment by AlexandreHiroyuki** [comment]
# Features and TODO

Reference: #3

**Comment by AlexandreHiroyuki** [comment]
# Front-end Design

> Figma is a great tool for collaboration between designers and developers or as a prototype tool for consistent long-term development.
>
> Even so, I chose to use a simpler tool (aka pen and paper) due to the short time available for development and the fact that my team has only one person working on this project.
>
> "Fail fast".
>
> If the project had more collaborators or time available, I would certainly use Figma.

## Design System (Improvised)

### Choosing Colors

I want the design to communicate technology, control over your business machinery/products, and a touch of professional/exclusive access.

With that in mind, my first color choice is blue.

Blue light inhibits the pineal gland from producing melatonin because sunlight is almost the only source of natural blue light. And melatonin is the hormone that regulates the day cycle in animals' bodies.
For humans, high melatonin means it is nighttime and, therefore time to sleep.
So blue is a nice color for keeping the user awake and attentive, making them feel more productive. Also, a higher contrast shade with the background can make reading and identifying objects (like buttons) less tiring for the eyes.

Now speaking about the AI feature, I wanted to communicate the tool in a more professional environment given that AI is the big trend. So I chose brown as a secondary color because it is more luxurious and serious.
Besides, blue and brown (a shade of orange) are complementary colors, meaning they harmonize well while expressing their opposition of contexts.
I used this site to calculate the exact code for the complementary color: https://www.canva.com/colors/color-wheel/

After picking the main colors of my project's brand, I used this tool to generate the shades needed for the design system: https://www.skeleton.dev/docs/generator

### Components

**Important**: Any React component used less than twice in different files should **not** be abstracted to avoid wasting time with overengineering.

Some style variables used only once were separated using the variable name to keep the tag utility more descriptive and intuitive.

## Home Page - Sign In / Sign Up

With the POC mentality, a big landing page to present the product is undesirable. It's better to focus on delivering the product's main features.

Therefore, the website outside the main product has to be only one page, preferably without scrolling to keep it simple and attractive to the eyes on the presentation. This means short texts to avoid distractions, layout, and colors that draw the user's attention and willingness to click the buttons to start using the product.

I want the user to experience the product as soon as possible, so the sign-up form does not have unnecessary fields.

Conclusion: a one-screen page, a grid of 2x2, 1st column of a concise presentation about the product (title and description), 2nd column with sign-in and sign-up form.

- Nice form feedback to the user.
  - Highlight the specific input bar that is given an error.
  - Small text below the input quickly explaining the error.
  - For a general form or API error, show a pop-up.

## Dashboard Page

- Shows a list of devices.
- Add mock device.
- Delete mock device.
- Read mock device data.
- Edit mock device.

Animations: https://webcode.tools/css-generator/keyframe-animation

### Future Updates

Due to a lack of time, some features should be postponed.

- Plot devices' data in graphs (https://www.highcharts.com).
- Connect a real device.

## AI Report Page

It was meant to look like a formal paper document, but with a modern pleasurable look.

## Request Handling

I prefer to use Axios instead of writing fetch manually because the syntax becomes leaner, it doesn't require manual serialization and parsing, it has built-in error handling, and it is widely compatible with older browsers.

**Comment by AlexandreHiroyuki** [comment]
# Backend

## Database

### Relational Schema

User Table

| Attribute | Data Type | Constraints | Description |
| -------- | ------- | -------- | ------- |
| `id` | `SERIAL` | Primary key, Auto-increment | Unique identifier for each user |
| `email` | `VARCHAR(255)` | Unique, Not null | Email address of the user |
| `password` | `VARCHAR(255)` | Not Null | Encrypted password of the user |
| `created_at` | `TIMESTAMP` | Default: `NOW()` | Timestamp when the user was created |

Device Table

| Attribute | Data Type | Constraints | Description |
| -------- | ------- | -------- | ------- |
| `id` | `SERIAL` | Primary key, Auto-increment | Unique identifier for each device |
| `user_id` | `INTEGER` | Not null | Unique identifier of the user(users table) that owns the device |
| `name` | `TEXT` | Not Null | Encrypted password of the user |
| `created_at` | `TIMESTAMP` | Default: `NOW()` | Timestamp when the user was created |

## API

I considered using Swagger to document the API. Still, the API is not the main product, so it's better to focus on delivering the main features and document only with classic markdown text.

HTTP Method | Path | Controller & Action | Description
-- | -- | -- | --
POST | /signup | userController.create | Registers a new user.
POST | /signin | userController.access | Authenticates a user and provides access.
POST | /device | deviceController.create | Adds a new device to the system.
GET | /device | deviceController.getList | Retrieves a list of all devices.
PUT | /device/:id | deviceController.update | Updates the details of a specific device.
DELETE | /device/:id | deviceController.delete | Deletes a specific device by ID.

**Comment by AlexandreHiroyuki** [comment]
After studying, it seems [sessions authentication](https://remix.run/docs/en/1.19.3/utils/sessions) may be more compatible with Remix SSR because of actions and loaders access. However, it is too late to change it now. At least I learned something.

**Comment by AlexandreHiroyuki** [comment]
Bug: React Aria modal doesn't react to `useState()`, and its default commands don't work.

**Comment by AlexandreHiroyuki** [comment]
Future TODO: handle API errors better.

https://stackoverflow.com/questions/44669508/whats-the-good-http-status-code-to-return-on-expired-password

**Comment by AlexandreHiroyuki** [comment]
A better way to run migrations on production:
https://dev.to/woovi/best-way-to-run-migrations-in-production-ehg
